spec/mongo/server/connection_spec.rb

```diff
@@ -400,6 +400,35 @@
       end
     end
 
+    context 'when a socket timeout occurs' do
+      let(:port) { 31104 }
+
+      let(:client) do
+        Mongo::Client.new([ "127.0.0.1:#{port}" ], database: "moped_test", socket_timeout: 1.5)
+      end
+
+      before do
+        start_mongo_server(port, "--setParameter enableTestCommands=1")
+        Process.detach(spawn("echo 'db.adminCommand({sleep: 1, w: true, secs: 5})' | mongo localhost:#{port} 2>&1 > /dev/null"))
+        sleep 1 # to sleep command on mongodb begins work
+      end
+
+      after do
+        stop_mongo_server(port)
+      end
+
+      it 'raises a operation timeout exception' do
+        time = Benchmark.realtime do
+          expect {
+            Timeout::timeout(3) do
+              client[:users].find("age" => { "$gte" => 65 }).first
+            end
+          }.to raise_exception("Took more than 1.5 seconds to receive data.")
+        end
+        expect(time).to be_within(0.2).of(1.5)
+      end
+    end
+
```

Comment:
I usually prefer not to check String literals in tests and I didn't like the idea of the test:

```ruby
        expect {
          Timeout::timeout(3) do
            client[authorized_collection.name].find("$where" => "sleep(10000) || true").first
          end
        }.to raise_exception(Timeout::Error)
```

to be true regardless of whether the socket timeout of 1.5 was actually applied. I know we would also have a test on the length of time anyway that would fail but I wanted to only test the length of time, not the error.

I don't feel that strongly on this point so we can change it to the following if you both want:

```ruby
      it 'raises a timeout when it expires' do
        start = Time.now
        expect {
          Timeout::timeout(3) do
            client[authorized_collection.name].find("$where" => "sleep(10000) || true").first
          end
        }.to raise_exception("Took more than 1.5 seconds to receive data.")
        end_time = Time.now
        expect(end_time - start).to be_within(0.2).of(1.5)
      end
    end
```